internal/controller/cilium_flow_manager.go

```diff
@@ -0,0 +1,228 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/cilium/cilium/api/v1/flow"
+	observer "github.com/cilium/cilium/api/v1/observer"
+	pb "github.com/illumio/cloud-operator/api/illumio/cloud/k8scluster/v1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// FlowManager holds a logger and Cilium client
+type FlowManager struct {
+	logger *zap.SugaredLogger
+	client observer.ObserverClient
+}
+
+// discoverHubbleRelayAddress uses a kubernetes clientset in order to discover the address of hubble-relay within kube-system.
+func discoverHubbleRelayAddress(ctx context.Context, logger *zap.SugaredLogger, clientset kubernetes.Interface) (string, error) {
+	service, err := clientset.CoreV1().Services("kube-system").Get(ctx, "hubble-relay", metav1.GetOptions{})
+	if err != nil {
+		logger.Errorw("Failed to get Hubble Relay service", "error", err)
+		return "", err
+	}
+
+	if len(service.Spec.Ports) == 0 {
+		logger.Errorw("Hubble Relay service has no ports", "error", err)
+		return "", errors.New("hubble relay service has no ports")
+	}
+
+	address := fmt.Sprintf("%s:%d", service.Spec.ClusterIP, service.Spec.Ports[0].Port)
+	return address, nil
+}
+
+// initFlowManager connects to hubble, sets up a observerClient and creates a new FlowManager object.
+func initFlowManager(ctx context.Context, logger *zap.SugaredLogger) (FlowManager, error) {
+	// Connect to Hubble
+	config, err := NewClientSet()
+	if err != nil {
+		logger.Errorw("Could not create a new client set", "error", err)
+		return FlowManager{}, err
+	}
+	hubbleAddress, err := discoverHubbleRelayAddress(ctx, logger, config)
+	if err != nil {
+		logger.Errorw("Cannot find hubble-relay address", "error", err)
+		return FlowManager{}, err
+	}
+	// Adjust this address if needed
+	conn, err := grpc.NewClient(hubbleAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		logger.Errorw("Failed to connect to Hubble", "error", err)
+		return FlowManager{}, err
+	}
+	hubbleClient := observer.NewObserverClient(conn)
+	return FlowManager{logger: logger, client: hubbleClient}, nil
```

Comment:
Don't log in this method.
Just return meaningful errors, and let the caller (`BootUpStreamAndReconnect`) log only once.
```suggestion
	config, err := NewClientSet()
	if err != nil {
		return nil, fmt.Errorf("failed to create new client set: %w", err)
	}
	hubbleAddress, err := discoverHubbleRelayAddress(ctx, logger, config)
	if err != nil {
		return nil, err
	}
	// Adjust this address if needed
	conn, err := grpc.NewClient(hubbleAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("failed to connect to Cilium Hubble Relay: %w", err)
	}
	hubbleClient := observer.NewObserverClient(conn)
	return &CiliumFlowCollector{logger: logger, client: hubbleClient}, nil
```

Reply:
Done